File: dart/test/generator_test.dart
import 'package:odoid/odoid.dart';
import 'package:test/test.dart';

void main() {
  group('OdoIDGenerator', () {
    test('default namespace is "default" and length is 6', () {
      final g = OdoIDGenerator();
      expect(g.namespace, equals('default'));
      expect(g.length, equals(6));
    });

    test('custom namespace and length are stored', () {
      final g = OdoIDGenerator(namespace: 'acme', length: 8);
      expect(g.namespace, equals('acme'));
      expect(g.length, equals(8));
    });

    for (final length in [6, 7, 8]) {
      test('capacity matches kMax[$length]', () {
        expect(OdoIDGenerator(length: length).capacity, equals(kMax[length]));
      });

      test('next() result id has correct length ($length)', () {
        expect(OdoIDGenerator(length: length).next().id.length, equals(length));
      });

      test('nextN() is in [0, capacity) for length $length', () {
        final g = OdoIDGenerator(length: length);
        for (var i = 0; i < 50; i++) {
          final n = g.nextN();
          expect(n, greaterThanOrEqualTo(0));
          expect(n, lessThan(g.capacity));
        }
      });
    }

    test('unsupported length throws UnsupportedLengthException', () {
      expect(() => OdoIDGenerator(length: 5),
          throwsA(isA<UnsupportedLengthException>()));
    });

    test('next() result has correct shape', () {
      final g = OdoIDGenerator(namespace: 'test', length: 6);
      final r = g.next();
      expect(r.id, isNotEmpty);
      expect(r.n, greaterThanOrEqualTo(0));
      expect(r.length, equals(6));
      expect(r.namespace, equals('test'));
    });

    test('next() id is uppercase alphanumeric', () {
      final g = OdoIDGenerator(length: 8);
      for (var i = 0; i < 50; i++) {
        final id = g.next().id;
        expect(id, equals(id.toUpperCase()),
            reason: 'ID not uppercase: $id');
        expect(RegExp(r'^[A-Z0-9]+$').hasMatch(id), isTrue,
            reason: 'ID contains non-alphanumeric: $id');
      }
    });

    test('excluded chars never appear in output', () {
      final g = OdoIDGenerator(length: 8);
      for (var i = 0; i < 50; i++) {
        final id = g.next().id;
        expect(id.contains('I'), isFalse, reason: '$id contains I');
        expect(id.contains('L'), isFalse, reason: '$id contains L');
        expect(id.contains('O'), isFalse, reason: '$id contains O');
      }
    });

    test('next().n matches decode(next().id)', () {
      final g = OdoIDGenerator(namespace: 'verify', length: 6);
      for (var i = 0; i < 20; i++) {
        final r = g.next();
        expect(OdoId.decode(r.id), equals(r.n));
      }
    });

    test('monotonic sequencing: same-tick produces distinct IDs', () {
      // epoch far in the future keeps tick == 0 for all calls
      final futureEpoch =
          DateTime.now().millisecondsSinceEpoch + 60000;
      final g = OdoIDGenerator(namespace: 'seq-test', epoch: futureEpoch);
      final seen = <String>{};
      for (var i = 0; i < 20; i++) {
        final id = g.next().id;
        expect(seen.contains(id), isFalse, reason: 'Duplicate ID: $id');
        seen.add(id);
      }
    });

    test('namespace isolation: different namespaces produce different IDs', () {
      final futureEpoch =
          DateTime.now().millisecondsSinceEpoch + 60000;
      final g1 = OdoIDGenerator(namespace: 'ns-a', length: 8, epoch: futureEpoch);
      final g2 = OdoIDGenerator(namespace: 'ns-b', length: 8, epoch: futureEpoch);
      expect(g1.next().id, isNot(equals(g2.next().id)));
    });

    test('proxy encode uses generator length', () {
      final g = OdoIDGenerator(length: 7);
      expect(g.encode(0).length, equals(7));
    });

    test('proxy decode returns correct n', () {
      final g = OdoIDGenerator(length: 6);
      final r = g.next();
      expect(g.decode(r.id), equals(r.n));
    });
  });
}

File: ts/test/generator.test.ts
import { describe, test, expect } from "vitest";
import { OdoIDGenerator, UnsupportedLengthError, decode, MAX } from "../src/index.js";

describe("OdoIDGenerator", () => {
  describe("construction", () => {
    test("creates with default options", () => {
      const g = new OdoIDGenerator();
      expect(g.namespace).toBe("default");
      expect(g.length).toBe(6);
    });

    test("creates with custom namespace and length", () => {
      const g = new OdoIDGenerator({ namespace: "acme", length: 8 });
      expect(g.namespace).toBe("acme");
      expect(g.length).toBe(8);
    });

    test("capacity matches MAX for the configured length", () => {
      expect(new OdoIDGenerator({ length: 6 }).capacity).toBe(MAX[6]);
      expect(new OdoIDGenerator({ length: 7 }).capacity).toBe(MAX[7]);
      expect(new OdoIDGenerator({ length: 8 }).capacity).toBe(MAX[8]);
    });

    test("unsupported length throws UnsupportedLengthError", () => {
      // @ts-expect-error intentionally unsupported length
      expect(() => new OdoIDGenerator({ length: 5 })).toThrowError(UnsupportedLengthError);
    });
  });

  describe("next()", () => {
    test("result has correct shape", () => {
      const g = new OdoIDGenerator({ namespace: "test", length: 6 });
      const result = g.next();
      expect(result).toHaveProperty("id");
      expect(result).toHaveProperty("n");
      expect(result).toHaveProperty("length");
      expect(result).toHaveProperty("namespace");
    });

    test("id has the correct length", () => {
      for (const length of [6, 7, 8] as const) {
        const g = new OdoIDGenerator({ length });
        expect(g.next().id).toHaveLength(length);
      }
    });

    test("id is uppercase alphanumeric with no excluded chars", () => {
      const g = new OdoIDGenerator({ length: 8 });
      for (let i = 0; i < 50; i++) {
        expect(g.next().id).toMatch(/^[0-9A-Z]{8}$/);
        expect(g.next().id).not.toMatch(/[ILO]/);
      }
    });

    test("n is always in [0, capacity)", () => {
      const g = new OdoIDGenerator({ length: 7 });
      for (let i = 0; i < 50; i++) {
        const { n } = g.next();
        expect(n).toBeGreaterThanOrEqual(0n);
        expect(n).toBeLessThan(MAX[7]);
      }
    });

    test("n matches decode(id)", () => {
      const g = new OdoIDGenerator({ namespace: "verify", length: 6 });
      for (let i = 0; i < 20; i++) {
        const { id, n } = g.next();
        expect(decode(id)).toBe(n);
      }
    });

    test("namespace and length fields match generator config", () => {
      const g = new OdoIDGenerator({ namespace: "ns1", length: 7 });
      const result = g.next();
      expect(result.namespace).toBe("ns1");
      expect(result.length).toBe(7);
    });
  });

  describe("encode() and decode() proxy methods", () => {
    test("generator.encode() uses the generator's length", () => {
      const g = new OdoIDGenerator({ length: 7 });
      expect(g.encode(0n)).toHaveLength(7);
    });

    test("generator.decode() returns the original n", () => {
      const g = new OdoIDGenerator({ length: 6 });
      const { id, n } = g.next();
      expect(g.decode(id)).toBe(n);
    });
  });

  describe("monotonic sequencing within a tick", () => {
    test("same-tick calls increment sequence and produce distinct IDs", () => {
      const epoch = Date.now() + 60_000; // future epoch — tick stays 0
      const g = new OdoIDGenerator({ namespace: "seq-test", epoch });

      const ids = new Set<string>();
      for (let i = 0; i < 20; i++) {
        ids.add(g.next().id);
      }

      // All IDs produced at the same tick (0) should be distinct
      expect(ids.size).toBe(20);
    });
  });

  describe("namespace isolation", () => {
    test("generators with different namespaces produce different IDs for the same tick", () => {
      const epoch = Date.now() + 60_000;
      const g1 = new OdoIDGenerator({ namespace: "ns-a", length: 8, epoch });
      const g2 = new OdoIDGenerator({ namespace: "ns-b", length: 8, epoch });

      const id1 = g1.next().id;
      const id2 = g2.next().id;

      expect(id1).not.toBe(id2);
    });
  });
});
